File: application/app/src/main/java/com/krs/community/fragments/BrowseByCityFragment.java
package com.krs.community.fragments;

import android.content.Context;
import android.os.AsyncTask;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.RecyclerView;

import com.ericliu.asyncexpandablelist.CollectionView;
import com.ericliu.asyncexpandablelist.async.AsyncExpandableListView;
import com.ericliu.asyncexpandablelist.async.AsyncExpandableListViewCallbacks;
import com.ericliu.asyncexpandablelist.async.AsyncHeaderViewHolder;
import com.krs.community.R;
import com.krs.community.activity.DashboardActivity;
import com.krs.community.model.City;
import com.krs.community.utils.Utility;
import com.krs.community.viewmodel.BrowseCityViewModel;
import com.krs.community.viewmodel.RegisterViewModel;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

import static android.content.ContentValues.TAG;

public class BrowseByCityFragment extends Fragment implements AsyncExpandableListViewCallbacks<String, City> {

    private AsyncExpandableListView<String, City> mAsyncExpandableListView;
    private CollectionView.Inventory<String, City> inventory;
    private BrowseCityViewModel browseCityViewModel;


    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {

        LayoutInflater layoutInflater = LayoutInflater.from(getActivity());

        View view = layoutInflater.inflate(R.layout.fragment_browse_city, container, false);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            Utility.changeStatusbarColor(getActivity(), R.color.colorBG, false);
        }

        browseCityViewModel = ViewModelProviders.of(this,factory).get(RegisterViewModel::class.java)
        browseCityViewModel.iBrow=this;

        ImageView iv_cancel = view.findViewById(R.id.iv_cancel);
        iv_cancel.setOnClickListener(v -> {
            Utility.movetoFragment(getActivity(), new DashboardFragment());
        });

        mAsyncExpandableListView = view.findViewById(R.id.asyncExpandableCollectionView);
        mAsyncExpandableListView.setCallbacks(this);

        inventory = new CollectionView.Inventory<>();

        String[] states = new String[]{"Gujarat", "Maharashtra", "Rajashtan", "Delhi", "Madhya Pradesh"};

        for (int i = 0; i < 5; i++) {
            CollectionView.InventoryGroup<String, City> group = inventory.newGroup(i); // groupOrdinal is the smallest, displayed first
            group.setHeaderItem(states[i]);
        }
        mAsyncExpandableListView.updateInventory(inventory);
        return view;
    }

    @Override
    public void onResume() {
        super.onResume();
        ((AppCompatActivity) getActivity()).getSupportActionBar().hide();
        DashboardActivity.spaceNavigationView.setVisibility(View.GONE);
    }

    @Override
    public void onStop() {
        super.onStop();
        ((AppCompatActivity) getActivity()).getSupportActionBar().show();
        DashboardActivity.spaceNavigationView.setVisibility(View.VISIBLE);
    }

    @Override
    public void onStartLoadingGroup(int groupOrdinal) {
        new LoadDataTask(groupOrdinal, mAsyncExpandableListView).executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
    }

    @Override
    public AsyncHeaderViewHolder newCollectionHeaderView(Context context, int groupOrdinal, ViewGroup parent) {

        View v = LayoutInflater.from(context).inflate(R.layout.header_row_item_async, parent, false);
        return new MyHeaderViewHolder(v, groupOrdinal, mAsyncExpandableListView);
    }

    @Override
    public RecyclerView.ViewHolder newCollectionItemView(Context context, int groupOrdinal, ViewGroup parent) {

        View v = LayoutInflater.from(context).inflate(R.layout.text_row_item_async, parent, false);

        return new NewsItemHolder(v);
    }

    @Override
    public void bindCollectionHeaderView(Context context, AsyncHeaderViewHolder holder, int groupOrdinal, String headerItem) {
        MyHeaderViewHolder myHeaderViewHolder = (MyHeaderViewHolder) holder;
        myHeaderViewHolder.getTextView().setText(headerItem);
    }

    @Override
    public void bindCollectionItemView(Context context, RecyclerView.ViewHolder holder, int i, City item) {
        NewsItemHolder newsItemHolder = (NewsItemHolder) holder;
        // newsItemHolder.getTextViewTitle().setText(item.getNewsTitle());
        newsItemHolder.getTextViewCity().setText(item.getCityName());

        if (item.getCityName().equalsIgnoreCase("other")) {
            newsItemHolder.getTextViewDevider().setVisibility(View.GONE);
        } else {
            newsItemHolder.getTextViewDevider().setVisibility(View.VISIBLE);
        }
    }

    private static class LoadDataTask extends AsyncTask<Void, Void, Void> {

        private final int mGroupOrdinal;
        private WeakReference<AsyncExpandableListView<String, City>> listviewRef = null;

        public LoadDataTask(int groupOrdinal, AsyncExpandableListView<String, City> listview) {
            mGroupOrdinal = groupOrdinal;
            listviewRef = new WeakReference<>(listview);
        }

        @Override
        protected Void doInBackground(Void... params) {
            try {
                Thread.sleep(1500);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            return null;
        }


        @Override
        protected void onPostExecute(Void aVoid) {
            List<City> items = new ArrayList<>();
            City news = new City();
            // news.setNewsTitle("Lawyers meet voluntary pro bono target for first time since 2013");
            news.setCityName("Ahmedabad");
            items.add(news);

            news = new City();
            //news.setNewsTitle("HSC 2016: 77000 students to sit first exams across NSW");
            news.setCityName("Gandhinagar");
            items.add(news);

            news = new City();
            //news.setNewsTitle("HSC 2016: 77000 students to sit first exams across NSW");
            news.setCityName("Other");
            items.add(news);

            if (listviewRef.get() != null) {
                listviewRef.get().onFinishLoadingGroup(mGroupOrdinal, items);
            }
        }

    }

    public class NewsItemHolder extends RecyclerView.ViewHolder {

        private final TextView tv_city;
        private final View view_devider;
        private final LinearLayout row_city;


        public NewsItemHolder(View v) {
            super(v);
            // Define click listener for the ViewHolder's View.
            v.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    Log.d(TAG, "Element " + getPosition() + " clicked.");
                }
            });

            tv_city = v.findViewById(R.id.tv_city);
            view_devider = v.findViewById(R.id.view_devider);
            row_city= v.findViewById(R.id.row_city);

            row_city.setOnClickListener(v1 -> Toast.makeText(getActivity(), ""+tv_city.getText(), Toast.LENGTH_SHORT).show());

        }

        /*public TextView getTextViewTitle() {
            return tvTitle;
        }*/
        public View getTextViewDevider() {
            return view_devider;
        }

        public TextView getTextViewCity() {
            return tv_city;
        }

    }

    public static class MyHeaderViewHolder extends AsyncHeaderViewHolder implements AsyncExpandableListView.OnGroupStateChangeListener {

        private final TextView textView;
        private final ProgressBar mProgressBar;
        private ImageView ivExpansionIndicator;

        public MyHeaderViewHolder(View v, int groupOrdinal, AsyncExpandableListView asyncExpandableListView) {
            super(v, groupOrdinal, asyncExpandableListView);
            textView = v.findViewById(R.id.title);
            mProgressBar = v.findViewById(R.id.progressBar);
            mProgressBar.getIndeterminateDrawable().setColorFilter(0xFFFFFFFF,android.graphics.PorterDuff.Mode.MULTIPLY);
            ivExpansionIndicator = v.findViewById(R.id.ivExpansionIndicator);
        }


        public TextView getTextView() {
            return textView;
        }


        @Override
        public void onGroupStartExpending() {
            mProgressBar.setVisibility(View.VISIBLE);
            ivExpansionIndicator.setVisibility(View.GONE);
        }

        @Override
        public void onGroupExpanded() {
            mProgressBar.setVisibility(View.GONE);
            ivExpansionIndicator.setVisibility(View.VISIBLE);
            ivExpansionIndicator.setImageResource(R.drawable.ic_arrow_up);
        }

        @Override
        public void onGroupCollapsed() {
            mProgressBar.setVisibility(View.GONE);
            ivExpansionIndicator.setVisibility(View.VISIBLE);
            ivExpansionIndicator.setImageResource(R.drawable.ic_arrow_down);
        }
    }
}

File: application/app/src/main/java/com/krs/community/fragments/BrowseByCityFragment.kt
package com.krs.community.fragments

import android.content.ContentValues.TAG
import android.content.Context
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.RecyclerView
import com.ericliu.asyncexpandablelist.CollectionView
import com.ericliu.asyncexpandablelist.async.AsyncExpandableListView
import com.ericliu.asyncexpandablelist.async.AsyncExpandableListViewCallbacks
import com.ericliu.asyncexpandablelist.async.AsyncHeaderViewHolder
import com.facebook.shimmer.ShimmerFrameLayout
import com.krs.community.R
import com.krs.community.activity.DashboardActivity
import com.krs.community.databinding.FragmentBrowseCityBinding
import com.krs.community.entities.City
import com.krs.community.utils.Coroutines
import com.krs.community.utils.Utility
import com.krs.community.viewmodel.BrowseCityViewModel
import com.krs.community.viewmodel.BrowseCityViewModelFactory
import kotlinx.android.synthetic.main.fragment_browse_city.view.*
import org.kodein.di.KodeinAware
import org.kodein.di.android.x.kodein
import org.kodein.di.generic.instance

class BrowseByCityFragment : Fragment(), AsyncExpandableListViewCallbacks<String, City>,KodeinAware {

    private lateinit var mAsyncExpandableListView: AsyncExpandableListView<String, City>
    private lateinit var shimmer_view_container: ShimmerFrameLayout

    private var inventory: CollectionView.Inventory<String, City>? = null
    private val factory: BrowseCityViewModelFactory by instance()
    internal var browseCityViewModel: BrowseCityViewModel? = null
    override val kodein by kodein()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            Utility.changeStatusbarColor(activity, R.color.colorBG, false)
        }

        val binding: FragmentBrowseCityBinding = DataBindingUtil.inflate(inflater, R.layout.fragment_browse_city, container, false)
        val view=  binding.root

        browseCityViewModel = ViewModelProviders.of(this,factory).get(BrowseCityViewModel::class.java)

        mAsyncExpandableListView = view.findViewById(R.id.asyncExpandableCollectionView)
        mAsyncExpandableListView.setCallbacks(this)

        inventory = CollectionView.Inventory()

        shimmer_view_container=view.findViewById(R.id.shimmer_view_container)
        shimmer_view_container.startShimmerAnimation()
        shimmer_view_container.visibility = View.VISIBLE

        getStatesFromDB()

        view.iv_cancel.setOnClickListener { v -> Utility.movetoFragment(activity, DashboardFragment()) }
        return view
    }

    private fun getStatesFromDB()=Coroutines.main{
        browseCityViewModel?.getStates()?.observeForever {
            for ((index, stateData) in it.withIndex()) {
                val group = inventory?.newGroup(index) //Integer.parseInt(stateData.id)// groupOrdinal is the smallest, displayed first
                group?.headerItem = stateData.name
            }
            shimmer_view_container.stopShimmerAnimation()
            shimmer_view_container.visibility = View.GONE
            mAsyncExpandableListView.updateInventory(inventory)
        }
    }

    private fun getCityFromDB(stateId:Int)=Coroutines.main{
        browseCityViewModel?.getCitiesByState(stateId)?.observeForever {
            val groupOrdinal=stateId-1
            mAsyncExpandableListView.onFinishLoadingGroup(groupOrdinal,it)
        }
    }


    override fun onResume() {
        super.onResume()
        (activity as AppCompatActivity).supportActionBar?.hide()
        DashboardActivity.spaceNavigationView.visibility = View.GONE
    }

    override fun onStop() {
        super.onStop()
        (activity as AppCompatActivity).supportActionBar?.show()
        DashboardActivity.spaceNavigationView.visibility = View.VISIBLE
    }

    override fun onStartLoadingGroup(groupOrdinal: Int) {
        getCityFromDB(groupOrdinal+1)
    }

    override fun newCollectionHeaderView(context: Context, groupOrdinal: Int, parent: ViewGroup): AsyncHeaderViewHolder {
        val v = LayoutInflater.from(context).inflate(R.layout.header_row_item_async, parent, false)
        return StateViewHolder(v, groupOrdinal, mAsyncExpandableListView)
    }

    override fun newCollectionItemView(context: Context, groupOrdinal: Int, parent: ViewGroup): RecyclerView.ViewHolder {
        val v = LayoutInflater.from(context).inflate(R.layout.text_row_item_async, parent, false)
        return CityItemHolder(v)
    }

    override fun bindCollectionHeaderView(context: Context, holder: AsyncHeaderViewHolder, groupOrdinal: Int, headerItem: String) {
        val stateViewHolder = holder as StateViewHolder
        stateViewHolder.textView.text = headerItem
    }

    override fun bindCollectionItemView(context: Context, holder: RecyclerView.ViewHolder, i: Int, item: City) {
        val cityItemHolder = holder as CityItemHolder
        cityItemHolder.textViewCity.text = item.name
        cityItemHolder.city_id=item.id.toString()
        if (item.name.equals("other", ignoreCase = true)) {
            cityItemHolder.textViewDevider.visibility = View.GONE
        } else {
            cityItemHolder.textViewDevider.visibility = View.VISIBLE
        }
    }

    inner class CityItemHolder internal constructor(v: View) : RecyclerView.ViewHolder(v) {

        internal val textViewCity: TextView
        internal val textViewDevider: View
        internal var city_id:String=""

        init {
            // Define click listener for the ViewHolder's View.
            v.setOnClickListener { Log.d(TAG, "Element $position clicked.") }

            textViewCity = v.findViewById(R.id.tv_city)
            textViewDevider= v.findViewById(R.id.view_devider)
            val row_city = v.findViewById<LinearLayout>(R.id.row_city)

            row_city.setOnClickListener { v1 ->
                val fragment=SearchCityResult()
                val mBundle = Bundle()
                mBundle.putString("city_name", textViewCity.text.toString())
                mBundle.putString("city_id", city_id)
                fragment.setArguments(mBundle)
                Utility.movetoFragment(activity, fragment)
             }
        }
    }

    class StateViewHolder internal constructor(v: View, groupOrdinal: Int, asyncExpandableListView: AsyncExpandableListView<*, *>) : AsyncHeaderViewHolder(v, groupOrdinal, asyncExpandableListView), AsyncExpandableListView.OnGroupStateChangeListener {

        internal val textView: TextView
        private val mProgressBar: ProgressBar
        private val ivExpansionIndicator: ImageView

        init {
            textView = v.findViewById(R.id.title)
            mProgressBar = v.findViewById(R.id.progressBar)
            mProgressBar.indeterminateDrawable.setColorFilter(-0x1, android.graphics.PorterDuff.Mode.MULTIPLY)
            ivExpansionIndicator = v.findViewById(R.id.ivExpansionIndicator)
        }


        override fun onGroupStartExpending() {
            mProgressBar.visibility = View.VISIBLE
            ivExpansionIndicator.visibility = View.GONE
        }

        override fun onGroupExpanded() {
            mProgressBar.visibility = View.GONE
            ivExpansionIndicator.visibility = View.VISIBLE
            ivExpansionIndicator.setImageResource(R.drawable.ic_arrow_up)
        }

        override fun onGroupCollapsed() {
            mProgressBar.visibility = View.GONE
            ivExpansionIndicator.visibility = View.VISIBLE
            ivExpansionIndicator.setImageResource(R.drawable.ic_arrow_down)
        }
    }
}
